File: titus/titus/lib1/spec.py
#!/usr/bin/env python

# Copyright (C) 2014  Open Data ("Open Data" refers to
# one or more of the following companies: Open Data Partners LLC,
# Open Data Research LLC, or Open Data Capital LLC.)
# 
# This file is part of Hadrian.
# 
# Licensed under the Hadrian Personal Use and Evaluation License (PUEL);
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://raw.githubusercontent.com/opendatagroup/hadrian/master/LICENSE
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import math

from titus.fcn import Fcn
from titus.fcn import LibFcn
from titus.signature import Sig
from titus.signature import Sigs
from titus.datatype import *
from titus.errors import *
import titus.P as P

provides = {}
def provide(fcn):
    provides[fcn.name] = fcn
prefix = "m.special."

class NChooseK(LibFcn):
    name = prefix + "nChooseK"
    sig = Sig([{"n": P.Int()}, {"k": P.Int()}], P.Int())
    def __call__(self, state, scope, paramTypes, n, k):
        try:
            return nChooseK(n,k)
        except:
            raise PFARuntimeException("domain error")
provide(NChooseK())

class LnBetaFcn(LibFcn):
    name = prefix + "lnBeta"
    sig = Sig([{"a": P.Double()}, {"b": P.Double()}], P.Double())
    def __call__(self, state, scope, paramTypes, a, b):
        try:
            return logBetaFunction(a,b)
        except:
            raise PFARuntimeException("domain error")
provide(LnBetaFcn())

class ErfFcn(LibFcn):
    name = prefix + "erf"
    sig = Sig([{"a": P.Double()}], P.Double())
    def __call__(self, state, scope, paramTypes, a):
        try:
            return math.erf(a)
        except:
            raise PFARuntimeException("domain error")
provide(ErfFcn())

class ErfcFcn(LibFcn):
    name = prefix + "erfc"
    sig = Sig([{"a": P.Double()}], P.Double())
    def __call__(self, state, scope, paramTypes, a):
        try:
            return math.erfc(a)
        except:
            raise PFARuntimeException("domain error")
provide(ErfcFcn())

class LnGammaFcn(LibFcn):
    name = prefix + "lnGamma"
    sig = Sig([{"a": P.Double()}], P.Double())
    def __call__(self, state, scope, paramTypes, a):
        if (a > 0):
            try:
                return math.lgamma(a)
            except:
                raise PFARuntimeException("domain error")
        else:
            raise PFARuntimeException("domain error")
provide(LnGammaFcn())

#class IncompleteBetaFcn(LibFcn):
#    name = prefix + "incompletebetafcn"
#    sig = Sig([{"x": P.Double()}, {"a": P.Double()}, {"b": P.Double()}], P.Double())
#    def __call__(self, state, scope, paramTypes, x, a, b):
#        return incompleteBetaFunction(x, a, b)
#provide(IncompleteBetaFcn())

#class IncompleteBetaFcnInv(LibFcn):
#    name = prefix + "incompletebetafcninv"
#    sig = Sig([{"p": P.Double()}, {"a": P.Double()}, {"b": P.Double()}], P.Double())
#    def __call__(self, state, scope, paramTypes, p, a, b):
#        return inverseIncompleteBetaFunction(p, a, b)
#provide(IncompleteBetaFcnInv())

# http://mathworld.wolfram.com/RegularizedGammaFunction.html
#class RegularizedGammaPFcn(LibFcn):
#    name = prefix + "regularizedgammapfcn"
#    sig = Sig([{"x": P.Double()}, {"a": P.Double()}], P.Double())
#    def __call__(self, state, scope, paramTypes, x, a):
#        return regularizedGammaP(a, x)
#provide(RegularizedGammaPFcn())

# http://mathworld.wolfram.com/RegularizedGammaFunction.html
#class RegularizedGammaQFcn(LibFcn):
#    name = prefix + "regularizedgammaqfcn"
#    sig = Sig([{"x": P.Double()}, {"a": P.Double()}], P.Double())
#    def __call__(self, state, scope, paramTypes, x, a):
#        return regularizedGammaQ(a, x)
#provide(RegularizedGammaQFcn())

#########################################################################################
### The actual functions ################################################################
#########################################################################################

def logBetaFunction(a, b):
    if (a <= 0.0) or (b <= 0.0):
        raise PFARuntimeException("domain error")
    else:
        return math.lgamma(a) + math.lgamma(b) - math.lgamma(a + b)

# numerical recipes in C
def incompleteBetaFunction(x,a,b):
    lbeta = math.lgamma(a + b) - math.lgamma(a) - math.lgamma(b) \
            + a * math.log(x) + b * math.log(1.0 - x)
    if (x < (a + 1)/(a + b + 2)):
        return math.exp(lbeta) * contFractionBeta(a,b,x)/a
    else:
        return 1 - math.exp(lbeta) * contFractionBeta(b,a,1.-x)/b

# numerical recipes in C
def inverseIncompleteBetaFunction(p, a, b):
    a1 = a - 1.0
    b1 = b - 1.0
    ERROR = 1.0e-10

    if (p <= 0.0):
        return 0.0
    elif (p >= 1.0):
        return 1.0;
    elif (a >= 1.0) and (b >= 1.0):
        if (p < 0.5):
            pp = p
        else:
            pp = 1.0 - p;
        t = math.sqrt(-2.0*math.log(pp))
        x = (2.30753+t*0.27061)/(1.0+t*(0.99229+t*0.04481)) - t
        if (p < 0.5):
            x = -x
        al = ((x*x)-3.0)/6.0;
        h = 2.0/(1.0/(2.0*a-1.0)+1.0/(2.0*b-1.0));
        w = (x*math.sqrt(al+h)/h)-(1.0/(2.0*b-1)-1.0/(2.0*a-1.0))*(al+5.0/6.0-2.0/(3.0*h))
        x = a/(a+b*math.exp(2.0*w))
    else:
        lna = math.log(a/(a+b))
        lnb = math.log(b/(a+b))
        t = math.exp(a*lna)/a
        u = math.exp(b*lnb)/b
        w = t + u
        if (p < t/w):
            x = math.pow(a*w*p,1.0/a)
        else:
            x = 1.0 - math.pow(b*w*(1.0-p), 1.0/b)
    afac = -math.lgamma(a)-math.lgamma(b)+math.lgamma(a+b)
    j = 0
    for i in range(10):
        if (0.0) or ( x == 1.0):
            return x;
        err = incompleteBetaFunction(x,a,b) - p
        t = math.exp(a1*math.log(x)+b1*math.log(1.0-x) + afac)
        u = err/t
        t = u/(1.0-0.5*min(1.0,u*(a1/x - b1/(1.0-x))))
        x -= t
        if (x <= 0.0):
            x = 0.5*(x + t)
        if (x >= 1.0):
            x = 0.5*(x + t + 1.0)
        if (math.fabs(t) < ERROR*x) and (j > 0):
            break
    return x

# numerical recipes in C
# used by incompleteBetaFcn and incompleteBetaFcnComp
def contFractionBeta(a,b,x, maxIter = 1000, epsilon=1e-15):
    bm = 1.0
    az = 1.0
    am = 1.0
    qab = a + b
    qap = a + 1.0
    qam = a - 1.0
    bz = 1.0 - qab*x/qap
    for i in range(0,maxIter):
        em = i + 1
        tem = em + em
        d = em*(b-em)*x/((qam+tem)*(a+tem))
        ap = az + d*am
        bp = bz+d*bm
        d = -(a+em)*(qab+em)*x/((qap+tem)*(a+tem))
        app = ap+d*az
        bpp = bp+d*bz
        aold = az
        am = ap/bpp
        bm = bp/bpp
        az = app/bpp
        bz = 1.0
        if (abs(az-aold) < (epsilon*abs(az))):
            return az
    raise PFARuntimeException("domain error")

# http://grepcode.com/file/repo1.maven.org/maven2/org.apache.commons/commons-math3/3.0/org/apache/commons/math3/special/Gamma.java#Gamma.regularizedGammaP%28double%2Cdouble%29
def regularizedGammaP(a, x):
    if (x < 0.0) or (a <= 0.0):
        raise PFARuntimeException("domain error")
    maxIter = 1000
    epsilon = 1e-15
    n  = 0.0
    an = 1.0/a
    total = an
    while (abs(an/total) > epsilon) and \
          (n < maxIter) and \
          (total < float("inf")):
        n = n + 1.0
        an = an * (x / (a + n))
        total = total + an
    if n >= maxIter:
        raise PFARuntimeException("domain error")
    elif total == float("inf"):
        ans = 1.0
    else:
        ans = math.exp(-x + (a*math.log(x)) - math.lgamma(a)) * total
    return ans

# http://www.johnkerl.org/python/sp_funcs_m.py.txt
def regularizedGammaQ(a, x):
    """Incomplete gamma function."""
    if (x < 0.0 or a <= 0.0):
        raise PFARuntimeException("domain error")
    if (x < a+1.0):
        return 1.0 - gser(a,x)[0]
    else:
        return gcf(a,x)[0]

# http://www.johnkerl.org/python/sp_funcs_m.py.txt
# used by incompleteGammaFcn
def gser(a, x, itmax=700, eps=3.0e-7):
    """Series approximation to the incomplete gamma function."""
    gln = math.lgamma(a)
    if (x < 0.0):
        raise PFARuntimeException("domain error")
    if (x == 0.0):
        return [0.0]
    ap = a
    total = 1.0 / a
    delta = total
    n = 1
    while n <= itmax:
        ap = ap + 1.0
        delta = delta * x / ap
        total = total + delta
        if (abs(delta) < abs(total)*eps):
            return [total * math.exp(-x + a*math.log(x) - gln), gln]
        n = n + 1
    raise PFARuntimeException("domain error")

# http://www.johnkerl.org/python/sp_funcs_m.py.txt
# used by incompleteGammaFcn
def gcf(a, x, itmax=200, eps=3.0e-7):
    """Continued fraction approximation of the incomplete gamma function."""
    gln = math.lgamma(a)
    gold = 0.0
    a0 = 1.0
    a1 = x
    b0 = 0.0
    b1 = 1.0
    fac = 1.0
    n = 1
    while n <= itmax:
        an = n
        ana = an - a
        a0 = (a1 + a0*ana)*fac
        b0 = (b1 + b0*ana)*fac
        anf = an*fac
        a1 = x*a0 + anf*a1
        b1 = x*b0 + anf*b1
        if (a1 != 0.0):
            fac = 1.0 / a1
            g = b1*fac
            if (abs((g-gold)/g) < eps):
                return (g*math.exp(-x+a*math.log(x)-gln), gln)
            gold = g
        n = n + 1
    raise PFARuntimeException("domain error")

def nChooseK(n,k):
    # is n an integer?
    nInt = (math.floor(n) == n)
    if (n <= k) or (k <= 0):
        raise PFARuntimeException("domain error")

    if (nInt) and (n < 0.0):
        b = pow(-1.0, k) * math.exp(math.lgamma(abs(n + k)) \
                                  - math.lgamma(k + 1.0)    \
                                  - math.lgamma(abs(n)))
        return round(b)
    if (n >= k):
        b = math.exp(math.lgamma(n + 1.0) - math.lgamma(k + 1.0) \
                   - math.lgamma(n - k + 1.0))
        return round(b)
    if not (nInt) and (n < k):
        b = (1.0/math.pi) * math.exp(math.lgamma(n + 1.0) \
                                   - math.lgamma(k + 1)   \
                                   + math.lgamma(k - n)   \
                   + math.log(math.sin(math.pi * (n - k + 1.0))))
        return round(b)
    return 0.0

File: hadrian/src/main/scala/com/opendatagroup/hadrian/lib/spec.scala
// Copyright (C) 2014  Open Data ("Open Data" refers to
// one or more of the following companies: Open Data Partners LLC,
// Open Data Research LLC, or Open Data Capital LLC.)
// 
// This file is part of Hadrian.
// 
// Licensed under the Hadrian Personal Use and Evaluation License (PUEL);
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://raw.githubusercontent.com/opendatagroup/hadrian/master/LICENSE
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.opendatagroup.hadrian.lib

import scala.annotation.tailrec
import scala.collection.immutable.ListMap

import org.apache.avro.Schema

import com.opendatagroup.hadrian.ast.LibFcn
import com.opendatagroup.hadrian.errors.PFARuntimeException
import com.opendatagroup.hadrian.errors.PFASemanticException
import com.opendatagroup.hadrian.jvmcompiler.JavaCode
import com.opendatagroup.hadrian.jvmcompiler.javaSchema
import com.opendatagroup.hadrian.jvmcompiler.javaType
import com.opendatagroup.hadrian.jvmcompiler.JVMNameMangle
import com.opendatagroup.hadrian.jvmcompiler.PFAEngineBase

import com.opendatagroup.hadrian.ast.AstContext
import com.opendatagroup.hadrian.ast.ExpressionContext
import com.opendatagroup.hadrian.ast.FcnDef
import com.opendatagroup.hadrian.ast.FcnRef

import com.opendatagroup.hadrian.data.PFAArray
import com.opendatagroup.hadrian.data.PFARecord

import com.opendatagroup.hadrian.signature.P
import com.opendatagroup.hadrian.signature.Sig
import com.opendatagroup.hadrian.signature.Signature
import com.opendatagroup.hadrian.signature.Sigs

import com.opendatagroup.hadrian.datatype.Type
import com.opendatagroup.hadrian.datatype.FcnType
import com.opendatagroup.hadrian.datatype.AvroType
import com.opendatagroup.hadrian.datatype.AvroNull
import com.opendatagroup.hadrian.datatype.AvroBoolean
import com.opendatagroup.hadrian.datatype.AvroInt
import com.opendatagroup.hadrian.datatype.AvroLong
import com.opendatagroup.hadrian.datatype.AvroFloat
import com.opendatagroup.hadrian.datatype.AvroDouble
import com.opendatagroup.hadrian.datatype.AvroBytes
import com.opendatagroup.hadrian.datatype.AvroFixed
import com.opendatagroup.hadrian.datatype.AvroString
import com.opendatagroup.hadrian.datatype.AvroEnum
import com.opendatagroup.hadrian.datatype.AvroArray
import com.opendatagroup.hadrian.datatype.AvroMap
import com.opendatagroup.hadrian.datatype.AvroRecord
import com.opendatagroup.hadrian.datatype.AvroField
import com.opendatagroup.hadrian.datatype.AvroUnion

import org.apache.commons.math3.special
import org.apache.commons.math3.special.Beta
import org.apache.commons.math3.distribution.BetaDistribution
import org.apache.commons.math3.special.Gamma
import org.apache.commons.math3.util.CombinatoricsUtils.binomialCoefficient

package object spec {
  private var fcns = ListMap[String, LibFcn]()
  def provides = fcns
  def provide(libFcn: LibFcn): Unit =
    fcns = fcns + Tuple2(libFcn.name, libFcn)

  val prefix = "m.special."

  ////   nChooseK (NChooseK)
  class NChooseK(val pos: Option[String] = None) extends LibFcn with Function2[Int, Int, Int] {
    def name = prefix + "nChooseK"
    def sig = Sig(List( "n" -> P.Int, "k" -> P.Int), P.Int)
    def doc =
      <doc>
        <desc> The number of ways to choose <p>k</p> elements from a set of <p>n</p> elements.</desc>
          <param name="n">Total number of elements.</param>
          <param name="k">Numer of elements chosen.</param>
        <ret>With <m>{"n"}</m> and <m>{"k"}</m>, this function evaluates the binomial coefficient.</ret> 
        <error code={s"${errcodeBase + 0}"}>Raises "domain error" if <m>{"k \\leq 0"}</m> or <m>{"k \\geq n"}</m>.</error>
      </doc>
    def errcodeBase = 36000

    def apply(n: Int, k: Int): Int =
      if (n <= k  ||  k <= 0)
        throw new PFARuntimeException("domain error", errcodeBase + 0, name, pos)
      else
        binomialCoefficient(n, k).toInt
  }
  provide(new NChooseK)

  ////   lnBeta (LnBeta)
  class LnBeta(val pos: Option[String] = None) extends LibFcn with Function2[Double, Double, Double] {
    def name = prefix + "lnBeta"
    def sig = Sig(List("a" -> P.Double, "b" -> P.Double), P.Double)
    def doc =
      <doc>
        <desc>Compute the beta function parameterized by <p>a</p> and <p>b</p>.</desc>
        <ret>With <m>{"a"}</m> and <m>{"b"}</m>, this function evaluates natural logarithm of the beta function. The beta function is <m>{"\\int_{0}^{1} t^{a - 1}(1 - t)^{b - 1} dt "}</m>.</ret> 
        <error code={s"${errcodeBase + 0}"}>Raises "domain error" if <m>{"a \\leq 0"}</m> or if <m>{"b \\leq 0"}</m>.</error>
      </doc>
    def errcodeBase = 36010

    def apply(a: Double, b: Double): Double =
      if (a <= 0.0  ||  b <= 0.0)
        throw new PFARuntimeException("domain error", errcodeBase + 0, name, pos)
      else
        Beta.logBeta(a,b)
  }
  provide(new LnBeta)

  ////   erf (Erf)
  class Erf(val pos: Option[String] = None) extends LibFcn with Function1[Double, Double] {
    def name = prefix + "erf"
    def sig = Sig(List("x" -> P.Double), P.Double)
    def doc =
    <doc>
    <desc>Return the error function of <p>x</p>.</desc>
    </doc>
    def errcodeBase = 36020
    def apply(x: Double): Double = special.Erf.erf(x)
  }
  provide(new Erf)

  // ////   erf (Erf)
  // class Erf(val pos: Option[String] = None) extends LibFcn with Function1[Double, Double] {
  //   val a1: Double =  0.254829592
  //   val a2: Double = -0.284496736
  //   val a3: Double =  1.421413741
  //   val a4: Double = -1.453152027
  //   val a5: Double =  1.061405429
  //   val  p: Double =  0.3275911

  //   def name = prefix + "special.erf"
  //   def sig = Sig(List("x" -> P.Double), P.Double)
  //   def doc =
  //     <doc>
  //       <desc>Return the error function of <p>x</p>.</desc>{wholeLine()}
  //     </doc>
  //   def errcodeBase = 36030
  //   def apply(x: Double): Double = {
  //     val sign = if (x < 0.0) -1.0 else 1.0
  //     val absx =  Math.abs(x)

  //     val t = 1.0/(1.0 + p*absx)
  //     val y = 1.0 - (((((a5*t + a4)*t + a3)*t + a2)*t + a1)*t * Math.exp(-x*x))
  //     sign*y
  //   }
  // }
  // provide(SpecialErf)

  ////   erfc (Erfc)
  class Erfc(val pos: Option[String] = None) extends LibFcn with Function1[Double, Double] {
    def name = prefix + "erfc"
    def sig = Sig(List("x" -> P.Double), P.Double)
    def doc =
      <doc>
        <desc>Return the complimentary error function of <p>x</p>.</desc>
      </doc>
    def errcodeBase = 36040
    def apply(x: Double): Double = special.Erf.erfc(x)
  }
  provide(new Erfc)

  ////   lnGamma (LnGamma)
  class LnGamma(val pos: Option[String] = None) extends LibFcn with Function1[Double, Double] {
    def name = prefix + "lnGamma"
    def sig = Sig(List("x" -> P.Double), P.Double)
    def doc =
      <doc>
        <desc>Return the natural log of the gamma function of <p>x</p>.</desc>
      </doc>
    def errcodeBase = 36050
    def apply(x: Double): Double = {
      if (x <= 0)
        throw new PFARuntimeException("domain error", errcodeBase + 0, name, pos)
      else
        Gamma.logGamma(x)
    }
  }
  provide(new LnGamma)
}

//////////////////////////////////////////// regularized beta function
//////   incompletebetafcn (IncompleteBetaFcn)
//  class IncompleteBetaFcn(val pos: Option[String] = None) extends LibFcn with Function3[Double, Double, Double, Double] {
//    def name = prefix + "incompletebetafcn"
//    def sig = Sig(List("x" -> P.Double, "a" -> P.Double, "b" -> P.Double), P.Double)
//    def doc =
//    <doc>
//    <desc>Compute the incomplete beta function at <p>x</p> parameterized by <p>a</p> and <p>b</p>.</desc>
//    <param name="a">First parameter.</param>
//    <param name="b">Second parameter.</param>
//
//    <ret>With <m>{"a"}</m> and <m>{"b"}</m>, this function evaluates the incomplete beta function.  
//    The beta function is <m>{"\\int_{0}^{x} t^{a - 1}(1 - t)^{b - 1} dt "}</m>.</ret> 
//    <error code={s"${errcodeBase + 0}"}>Raises "domain error" if <m>{"a \\leq 0"}</m> or, <m>{"b \\leq 0"}</m> or if <m>{"x \\leq 0"}</m>.</error>
//    </doc>
//    def errcodeBase = 36060
//
//    def apply(x: Double, a: Double, b: Double): Double =
//      if (a <= 0.0 || b <= 0.0 || x <= 0.0){
//        throw new PFARuntimeException("domain error", errcodeBase + 0, name, pos)
//      }else {
//        return new BetaDistribution(a,b).cumulativeProbability(x)
//      }
//  }
//  provide(new IncompleteBetaFcn)
//
//
//////////////////////////////////////////// inverse incomplete beta function
//////   incompletebetafcninv (IncompleteBetaFcninv)
//  class IncompleteBetaFcnInv(val pos: Option[String] = None) extends LibFcn with Function3[Double, Double, Double, Double] {
//    def name = prefix + "incompletebetafcninv"
//    def sig = Sig(List("x" -> P.Double, "a" -> P.Double, "b" -> P.Double), P.Double)
//    def doc =
//    <doc>
//    <desc>Compute the inverse of the incomplete beta function at <p>x</p> parameterized by <p>a</p> and <p>b</p>.</desc>
//    <param name="a">First parameter.</param>
//    <param name="b">Second parameter.</param>
//
//    <ret>With <m>{"a"}</m> and <m>{"b"}</m>, this function evaluates the incomplete beta function.  
//    The beta function is <m>{"\\int_{0}^{x} t^{a - 1}(1 - t)^{b - 1} dt "}</m>.</ret> 
//    <error code={s"${errcodeBase + 0}"}>Raises "domain error" if <m>{"a \\leq 0"}</m> or, <m>{"b \\leq 0"}</m> or if <m>{"x \\leq 0"}</m>.</error>
//    </doc>
//    def errcodeBase = 36070
//
//    def apply(x: Double, a: Double, b: Double): Double =
//      if (a <= 0.0 || b <= 0.0 || x <= 0.0){
//        throw new PFARuntimeException("domain error", errcodeBase + 0, name, pos)
//      }else {
//        return new BetaDistribution(a,b).inverseCumulativeProbability(x)
//      }
//  }
//  provide(IncompleteBetaFcnInv)
//
//
//////////////////////////////////////////// regularized gamma Q
//////   regularizedgammaqfcn (RegularizedGammaQFcn)
//  object RegularizedGammaQFcn extends LibFcn with Function2[Double, Double, Double] {
//    val name = prefix + "regularizedgammaqfcn"
//    val sig = Sig(List( "s" -> P.Double, "x" -> P.Double), P.Double)
//    val doc =
//    <doc>
//    <desc>Compute the regularized gamma Q parameterized by <p>s</p> and <p>x</p>.</desc>
//    <param name="s">First parameter.</param>
//    <param name="x">Second parameter.</param>
//
//    <ret>With <m>{"s"}</m> and <m>{"x"}</m>, this function evaluates the regularized gamma Q function, <m>{"\\frac{\\gamma(s,x)}{\\Gamma(s)}"}</m>, where <m>{"\\gamma(s,x)"}</m> is the lower incomplete gamma function and <m>{"\\Gamma(s)"}</m> is the gamma function.</ret> 
//    <error code={s"${errcodeBase + 0}"}>Raises "domain error" if <m>{"s"}</m> is zero or an integer less than zero, or if <m>{"x \\leq 0"}</m>.</error>
//    </doc>
//    def errcodeBase = 36080
//
//    def apply(s: Double, x: Double): Double =
//      if (x <= 0.0 || s == 0.0){
//        throw new PFARuntimeException("domain error", errcodeBase + 0, name, pos)
//      }else if (s < 0.0 && Math.floor(s) == s){
//        throw new PFARuntimeException("domain error", errcodeBase + 0, name, pos)
//      }else {
//        return Gamma.regularizedGammaQ(s, x)
//      }
//  }
//  provide(RegularizedGammaQFcn)
//
//
//////////////////////////////////////////// regularized gamma P
//////   regularizedgammapfcn (RegularizedGammaPFcn)
//  object RegularizedGammaPFcn extends LibFcn with Function2[Double, Double, Double] {
//    val name = prefix + "regularizedgammapfcn"
//    val sig = Sig(List( "s" -> P.Double, "x" -> P.Double), P.Double)
//    val doc =
//    <doc>
//    <desc>Compute the regularized gamma P parameterized by <p>s</p> and <p>x</p>.</desc>
//    <param name="s">First parameter.</param>
//    <param name="x">Second parameter.</param>
//
//    <ret>With <m>{"s"}</m> and <m>{"x"}</m>, this function evaluates the regularized gamma P function, <m>{"\\frac{\\Gamma(s,x)}{\\Gamma(s)}"}</m>, where <m>{"\\Gamma(s,x)"}</m> is the upper incomplete gamma function and <m>{"\\Gamma(s)"}</m> is the gamma function.</ret> 
//    <error code={s"${errcodeBase + 0}"}>Raises "domain error" if <m>{"s"}</m> is zero or an integer less than zero, or if <m>{"x \\leq 0"}</m>.</error>
//    </doc>
//    def errcodeBase = 36090
//
//    def apply(s: Double, x: Double): Double =
//      if (x <= 0.0 || s == 0.0){
//        throw new PFARuntimeException("domain error", errcodeBase + 0, name, pos)
//      }else if(s < 0.0 && Math.floor(s) == s){
//        throw new PFARuntimeException("domain error", errcodeBase + 0, name, pos)
//      }else {
//        return Gamma.regularizedGammaP(s, x)
//      }
//  }
//  provide(RegularizedGammaPFcn)
//
//
//
